File: core/src/main/java/com/github/xsi640/qinquery/core/visitor/Visitor.java
package com.github.xsi640.qinquery.core.visitor;

import com.github.xsi640.qinquery.core.ast.*;

/**
 * @author SuYang
 */
public interface Visitor<C> {
    void onParam(ParamExpression expr, C context);

    void onFunc(FuncExpression expr, C context);

    void onValue(ValueExpression expr, C context);

    void onOperator(OperatorExpression expor, C context);

    void onSymbol(SymbolExpression expr, C context);
}

File: query-plus-core/src/main/kotlin/com/github/xsi640/queryplus/core/visitor/Visitor.kt
package com.github.xsi640.queryplus.core.visitor

import com.github.xsi640.queryplus.core.ast.*

interface Visitor<C, out R> {
    fun onLiteral(context: C, expr: LiteralExpression): R
    fun onParameter(context: C, expr: ParameterExpression): R
    fun onOperator(context: C, expr: OperatorExpression): R
    fun onEmpty(context: C, expr: EmptyExpression): R
    fun onQuery(context: C, expr: QueryExpression): R
}